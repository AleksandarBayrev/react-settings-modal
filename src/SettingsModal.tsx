import React from 'react';
import { sessionStorageKey } from './constants';

export type SettingsModalProps = {
    isAPIEnabled: boolean;
}

export type SettingsModalState = {
    isAPIEnabled: boolean;
}

export class SettingsModal extends React.Component<SettingsModalProps, SettingsModalState> {
    private checkboxRef: React.LegacyRef<HTMLInputElement> | null = null;
    constructor(props: SettingsModalProps) {
        super(props);
        this.checkboxRef = React.createRef();
        this.state = {
            isAPIEnabled: props.isAPIEnabled
        };
    }

    private onSettingChange(settingName: keyof SettingsModalState, settingValue: any, checkboxRef: React.LegacyRef<HTMLInputElement> | null = null) {
        this.setState({
            [settingName]: settingValue
        });
        setTimeout(() => {
            window.sessionStorage.setItem(sessionStorageKey, JSON.stringify(this.state));
            if (checkboxRef) {
                (checkboxRef as any).checked = Boolean(settingValue)
            }
        });
    }

    render(): React.ReactNode {
        return (
            <div className='settings-modal-container'>
                <div className='settings-modal-api-enabled'>
                    <div className='checkbox'>
                        <input 
                            ref={this.checkboxRef}
                            type='checkbox'
                            onChange={(e) => this.onSettingChange('isAPIEnabled', !this.state.isAPIEnabled)}
                            checked={this.state.isAPIEnabled}
                        />
                    </div>
                    <div
                        className='label'
                        onClick={() => this.onSettingChange('isAPIEnabled', !this.state.isAPIEnabled, this.checkboxRef)}
                    >Is API Enabled</div>
                </div>
            </div>
        )
    }
}