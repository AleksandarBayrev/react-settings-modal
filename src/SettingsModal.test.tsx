import React from 'react';
import renderer from 'react-test-renderer';
import { SettingsModal } from './SettingsModal';

describe('SettingsModal', () => {
    it('renders with settings on', () => {
        const component = renderer.create(<SettingsModal isAPIEnabled={true} />).toJSON();
        expect(component).toMatchSnapshot();
    });
    it('renders with settings off', () => {
        const component = renderer.create(<SettingsModal isAPIEnabled={false} />).toJSON();
        expect(component).toMatchSnapshot();
    });
})