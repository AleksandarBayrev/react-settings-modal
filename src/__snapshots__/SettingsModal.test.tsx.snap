// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsModal renders with settings off 1`] = `
<div
  className="settings-modal-container"
>
  <div
    className="settings-modal-api-enabled"
  >
    <div
      className="checkbox"
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
    </div>
    <div
      className="label"
      onClick={[Function]}
    >
      Is API Enabled
    </div>
  </div>
</div>
`;

exports[`SettingsModal renders with settings on 1`] = `
<div
  className="settings-modal-container"
>
  <div
    className="settings-modal-api-enabled"
  >
    <div
      className="checkbox"
    >
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
      />
    </div>
    <div
      className="label"
      onClick={[Function]}
    >
      Is API Enabled
    </div>
  </div>
</div>
`;
